File: frontend/src/api/orderStatus.ts
/**
 * Order Status API Client
 */

import apiClient from './index'
import type { OrderStatus, PaginatedResponse } from '@/types/orders'

export const orderStatusApi = {
  /**
   * List all order statuses
   */
  list() {
    return apiClient.get<PaginatedResponse<OrderStatus>>('/order-statuses/')
  },

  /**
   * Get single order status
   */
  get(id: number) {
    return apiClient.get<OrderStatus>(`/order-statuses/${id}/`)
  },

  /**
   * Get only active statuses
   */
  getActive() {
    return apiClient.get<PaginatedResponse<OrderStatus>>('/order-statuses/?is_active=true')
  },

  /**
   * Get next allowed statuses from current status
   */
  getNextStatuses(statusId: number) {
    return apiClient.get<PaginatedResponse<OrderStatus>>(`/order-statuses/${statusId}/next_statuses/`)
  },

  /**
   * Get complete workflow information
   */
  getWorkflow() {
    return apiClient.get<Record<string, unknown>>('/order-statuses/workflow/')
  },

  /**
   * Create new status (admin only)
   */
  create(data: Omit<OrderStatus, 'id'>) {
    return apiClient.post<OrderStatus>('/order-statuses/', data)
  },

  /**
   * Update status (admin only)
   */
  update(id: number, data: Partial<Omit<OrderStatus, 'id'>>) {
    return apiClient.patch<OrderStatus>(`/order-statuses/${id}/`, data)
  },

  /**
   * Delete status (admin only)
   */
  delete(id: number) {
    return apiClient.delete(`/order-statuses/${id}/`)
  }
}

File: backend/orders/api/viewsets/order_status.py
"""
Order Status ViewSet
"""

from django_filters.rest_framework import DjangoFilterBackend
from rest_framework import permissions, viewsets
from rest_framework.decorators import action
from rest_framework.filters import OrderingFilter, SearchFilter
from rest_framework.response import Response

from jf_manager_backend.permissions import DepartmentRoleModelPermissions, OrgWideWritePermission
from orders.api.filters import OrderStatusFilter
from orders.api.serializers import OrderStatusSerializer
from orders.models import OrderStatus


class OrderStatusViewSet(viewsets.ModelViewSet):
    """
    ViewSet for Order Status management

    Provides CRUD operations and workflow information
    """

    queryset = OrderStatus.objects.all()
    serializer_class = OrderStatusSerializer
    permission_classes = [permissions.IsAuthenticated, DepartmentRoleModelPermissions, OrgWideWritePermission]
    filter_backends = [DjangoFilterBackend, SearchFilter, OrderingFilter]
    filterset_class = OrderStatusFilter
    search_fields = ["name", "code", "description"]
    ordering_fields = ["sort_order", "name", "code"]
    ordering = ["sort_order", "name"]

    @action(detail=False, methods=["get"])
    def active(self, request):
        """Get only active statuses"""
        statuses = self.queryset.filter(is_active=True).order_by("sort_order")
        serializer = self.get_serializer(statuses, many=True)
        return Response(serializer.data)

    @action(detail=True, methods=["get"])
    def next_statuses(self, request, pk=None):
        """Get allowed next statuses from this status"""
        status = self.get_object()

        # Get statuses with higher sort_order
        next_statuses = OrderStatus.objects.filter(is_active=True, sort_order__gt=status.sort_order).order_by(
            "sort_order"
        )[:3]

        serializer = self.get_serializer(next_statuses, many=True)
        return Response(serializer.data)

    @action(detail=False, methods=["get"])
    def workflow(self, request):
        """Get complete workflow information"""
        from orders.notifications import OrderWorkflowService

        workflow = OrderWorkflowService.get_workflow_graph()
        return Response(workflow)
